src/Common/tests/Collections/IListTest.cs

```diff
@@ -0,0 +1,1692 @@
+                        () => list[0] = items[1]);
+                    CollectionAssert.Equal(items, list);
+                }
+                else
+                {
+                    var items = GenerateItems(32);
+                    var list = GetList(items);
+                    for (int i = 0; i < items.Length; i++)
+                    {
+                        list[i] = items[items.Length - i - 1];
+                    }
+                    Array.Reverse(items);
+                    CollectionAssert.Equal(items, list);
+                }
+
+                {
+                    var items = GenerateItems(32);
+                    var origItems = (object[]) items.Clone();
+                    var list = GetList(items);
+
+                    // verify setting every item
+                    for (int i = 0; i < items.Length; i++)
+                    {
+                        if (i < items.Length/2)
+                        {
+                            list[items.Length - i - 1] = GenerateItem();
+                        }
+                        list[i] = items[items.Length - i - 1];
+                    }
+
+                    Array.Reverse(items);
+                    CollectionAssert.Equal(items, list);
+
+                    // verify setting items back to the original
+                    items = origItems;
+                    if (ItemsMustBeUnique)
+                    {
+                        for (int i = 0; i < items.Length; i++)
+                            list[i] = GenerateItem();
+                    }
+
+                    for (int i = 0; i < items.Length; i++)
+                    {
+                        list[i] = items[i];
+                    }
+
+                    CollectionAssert.Equal(items, list);
+                }
+            }
+        }
+
+        [Fact]
+        public void AddFixedSizeOrReadonlyThrows()
+        {
+            if (ExpectedIsFixedSize || ExpectedIsReadOnly)
+            {
+                var items = GenerateItems(10);
+                var list = GetList(items);
+                Assert.Throws<NotSupportedException>(
+                    () => list.Add(GenerateItem()));
+                CollectionAssert.Equal(items, list);
+            }
+        }
+
+        [Fact]
+        public void AddWithNullValueInMiddle()
+        {
+            if (!ExpectedIsFixedSize && !ExpectedIsReadOnly)
+            {
+                object[] items;
+                var list = GetList(null);
+                var tempItems = GenerateItems(32);
+                for (int i = 0; i < tempItems.Length/2; i++)
+                {
+                    list.Add(tempItems[i]);
+                }
+
+                if (ItemsMustBeNonNull)
+                {
+                    items = tempItems;
+                    Assert.Throws<ArgumentNullException>(
+                        () => list.Add(null));
+                }
+                else
+                {
+                    items = new object[33];
```

Comment:
All of the numbers past this point should be derived from one original number.  (Precomputed method-const values, or just inline math)